sys/include/storage/flash_sim.h

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2015 Lucas Jenß
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_storage
+ * @brief
+ * @{
+ *
+ * @brief       Simulates flash memory for easier testing of storage subsystems
+ * @author      Lucas Jenß <lucas@x3ro.de>
+ */
+
+#ifndef _FS_FLASH_SIM_H
+#define _FS_FLASH_SIM_H
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Enumerates all error codes returned by the flash simulation layer
+ */
+typedef enum flash_sim_error {
+    E_FS_SUCCESS = 0,            //!< Operation was successful
+    E_FS_INVALID_WRITE = 1,      //!< Trying to set a bit from 0 -> 1 w/ a program operation
+    E_FS_INVALID_PARAMS = 2,     //!< Invalid parameters were passed to the function
+    E_FS_OUT_OF_RANGE = 3,       //!< The operation was out of range
+    E_FS_FILE_ERROR = 4,         //!< Error operating on the file simulating the flash device
+    E_FS_NOT_INITIALIZED = 5     //!< Flash simulation layer was not initialized prior to operation
+} flash_sim_error_t;
```

Comment:
As far as I see everyone of these can be expressed through an errno.

```
E_FS_SUCCESS         = 0
E_FS_INVALID_WRITE   = -EIO    /* Input/output error */
E_FS_INVALID_PARAMS  = -EINVAL /* Invalid argument */
E_FS_OUT_OF_RANGE    = -EFAULT /* Bad address */
E_FS_FILE_ERROR      = /* multiple errnos related to files, depending on what you want, see errno(3) */
E_FS_NOT_INITIALIZED = -ENODEV /* No such device */
```

Making up new error classes for modules is not very meaningful or helpful IMHO, since:
1) most users don't check them anyways (only *IF* there was an error)
2) it makes passing errors up more complicated, since modules passing up the error have to first check what it was and then use their respective error code
3) it decouples documentation of the error case from the actual place where it happens. Your code already just states `E_SUCCESS or another ::flash_sim_error_t`. That makes it really hard for testers to say if a function can error on `E_FS_INVALID_WRITE`, `E_FS_OUT_OF_RANGE`, etc.